File: app/src/main/java/se/gustavkarlsson/skylight/android/background/providers/impl/DebugAuroraReportProvider.java
package se.gustavkarlsson.skylight.android.background.providers.impl;

import android.location.Address;
import android.support.annotation.NonNull;

import java.util.Locale;

import se.gustavkarlsson.skylight.android.background.providers.AuroraReportProvider;
import se.gustavkarlsson.skylight.android.models.AuroraFactors;
import se.gustavkarlsson.skylight.android.models.AuroraReport;
import se.gustavkarlsson.skylight.android.models.factors.Darkness;
import se.gustavkarlsson.skylight.android.models.factors.GeomagActivity;
import se.gustavkarlsson.skylight.android.models.factors.GeomagLocation;
import se.gustavkarlsson.skylight.android.models.factors.Visibility;
import se.gustavkarlsson.skylight.android.settings.DebugSettings;

public class DebugAuroraReportProvider implements AuroraReportProvider {

	private final DebugSettings debugSettings;

	public DebugAuroraReportProvider(DebugSettings debugSettings) {
		this.debugSettings = debugSettings;
	}

	@Override
	public AuroraReport getReport(long timeoutMillis) {
		Address location = new Address(Locale.ENGLISH);
		AuroraFactors auroraFactors = createAuroraFactors();
		return new AuroraReport(System.currentTimeMillis(), location, auroraFactors);
	}

	@NonNull
	private AuroraFactors createAuroraFactors() {
		GeomagActivity geomagActivity = new GeomagActivity(debugSettings.getKpIndex());
		GeomagLocation geomagLocation = new GeomagLocation(debugSettings.getGeomagLatitude());
		Darkness darkness = new Darkness(debugSettings.getSunZenithAngle());
		Visibility visibility = new Visibility(debugSettings.getCloudPercentage());
		return new AuroraFactors(
				geomagActivity,
				geomagLocation,
				darkness,
				visibility);
	}
}

File: app/src/debug/kotlin/se/gustavkarlsson/skylight/android/services_impl/providers/DebugAuroraReportProvider.kt
package se.gustavkarlsson.skylight.android.services_impl.providers

import io.reactivex.Single
import se.gustavkarlsson.skylight.android.entities.*
import se.gustavkarlsson.skylight.android.services.DebugSettings
import se.gustavkarlsson.skylight.android.services.providers.AuroraReportProvider
import se.gustavkarlsson.skylight.android.services.providers.TimeProvider
import java.util.concurrent.TimeUnit

class DebugAuroraReportProvider(
	private val realProvider: AuroraReportProvider,
	private val debugSettings: DebugSettings,
	private val timeProvider: TimeProvider
) : AuroraReportProvider {

	override fun get(): Single<AuroraReport> {
		return Single.fromCallable {
			if (debugSettings.overrideValues) {
				val auroraFactors = createDebugFactors()
				val locationName = "Fake Location"
				val timestamp = timeProvider.getTime().blockingGet()
				Single.just(AuroraReport(timestamp, locationName, auroraFactors))
					.delay(5, TimeUnit.SECONDS)
			} else {
				realProvider.get()
			}
		}.flatMap { it }

	}

	private fun createDebugFactors(): AuroraFactors {
		val kpIndex = KpIndex(debugSettings.kpIndex)
		val geomagLocation = GeomagLocation(debugSettings.geomagLatitude)
		val darkness = Darkness(debugSettings.sunZenithAngle)
		val visibility = Visibility(debugSettings.cloudPercentage)
		return AuroraFactors(kpIndex, geomagLocation, darkness, visibility)
	}
}
